### ISSUE TITLE
MySQL seem to be broken on develop branch

### ISSUE THREAD

**Issue by madmath03** [ether/etherpad-lite#3805]
Seems like there is an issue with MySQL on develop branch:
```
etherpad       | /opt/etherpad-lite/src/node_modules/mysql/lib/protocol/Parser.js:437
etherpad       |       throw err; // Rethrow non-MySQL errors
etherpad       |       ^
etherpad       | 
etherpad       | TypeError: Cannot read property 'character_set_name' of undefined
etherpad       |     at Query.<anonymous> (/opt/etherpad-lite/src/node_modules/ueberdb2/mysql_db.js:86:19)
etherpad       |     at Query.<anonymous> (/opt/etherpad-lite/src/node_modules/mysql/lib/Connection.js:526:10)
etherpad       |     at Query._callback (/opt/etherpad-lite/src/node_modules/mysql/lib/Connection.js:488:16)
etherpad       |     at Query.Sequence.end (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/sequences/Sequence.js:83:24)
etherpad       |     at Query._handleFinalResultPacket (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/sequences/Query.js:149:8)
etherpad       |     at Query.EofPacket (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/sequences/Query.js:133:8)
etherpad       |     at Protocol._parsePacket (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/Protocol.js:291:23)
etherpad       |     at Parser._parsePacket (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/Parser.js:433:10)
etherpad       |     at Parser.write (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/Parser.js:43:10)
etherpad       |     at Protocol.write (/opt/etherpad-lite/src/node_modules/mysql/lib/protocol/Protocol.js:38:16)
```

Got this in Travis-CI dockerfile test: https://travis-ci.org/github/Monogramm/docker-etherpad/jobs/669401045#L977

For info, this is custom dockerfile, but highly inspired from the official one from this repository, and the same configuration applied on 1.8.0 works.

**Comment by muxator** [comment]
@JohnMcLear, this is yours :)

**Comment by JohnMcLear** [comment]
+1 will look at this today

**Comment by JohnMcLear** [comment]
I get error

```
/home/jose/johnmcether/src/node_modules/ueberdb2/mysql_db.js:73
    if (result[0].DEFAULT_CHARACTER_SET_NAME !== db.charset){
```
If the password is wrong in ``settings.json`` -- that needs fixing too :)

**Comment by JohnMcLear** [comment]
```
  var tableCharSet = "SELECT CCSA.character_set_name FROM information_schema.`TABLES` T,information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHER$
  db.query(tableCharSet,function(err, result, tf){
    if (result[0].character_set_name !== db.charset){
      console.error("table is not configured with charset "+db.charset+" -- This may lead to crashes when certain characters are pasted in pads"); // offending lines..
      console.log(result[0], db.charset);
    };
  });
```

**Comment by JohnMcLear** [comment]
Should be fixed in https://github.com/ether/etherpad-lite/commit/4db91cb2ed1a547362144a82e59ac745a8cd79cf

TLDR; I was doing the query before the callback meaning sometimes MySQL wasn't available.  Timing is now right and we handle the error more gracefully if the charset is not defined.

I'm not a docker user or even familiar but I don't see anything in our docker file which specifies MySQL.  It's possibly my code did exactly what it's supposed to do and that's catch bad MySQL bad configs (albeit in a bad way).

Do you want to paste the commands you are using for MySQL config / setup and let me review them?

**Comment by madmath03** [comment]
> Do you want to paste the commands you are using for MySQL config / setup and let me review them?

There isn't anything particular done to setup MySQL/MariaDB because the official containers do it themselves.
Just setting the env var will automatically create the database and user:
```yml
  etherpad_db:
    image: mariadb:latest
    container_name: etherpad_db
    restart: always
    command: --character_set_client=utf8 --character-set-server=utf8mb4 --collation-server=utf8mb4_unicode_ci --character-set-client-handshake=FALSE
    #ports:
    #  - 3306:3306
    volumes:
      - /srv/etherpad/db:/var/lib/mysql
      - /etc/localtime:/etc/localtime:ro
      - /etc/timezone:/etc/timezone:ro
    environment:
      - MYSQL_ROOT_PASSWORD=mysqlrootpassword
      - MYSQL_DATABASE=${ETHERPAD_DB_NAME}
      - MYSQL_USER=${ETHERPAD_DB_USER}
      - MYSQL_PASSWORD=${ETHERPAD_DB_PASSWORD}
```

Etherpad-lite DB setup is then just done using the regular env var set in [settings.json](https://github.com/Monogramm/docker-etherpad/blob/master/template/settings.json#L141):
```yml
  etherpad:
    # ...
    environment:
      - TITLE=Test Etherpad
      # Database configuration
      - DB_TYPE=mysql
      - DB_HOST=etherpad_db # same as mysql container name
      - DB_PORT=3306
      - DB_NAME=${ETHERPAD_DB_NAME}
      - DB_USER=${ETHERPAD_DB_USER}
      - DB_PASS=${ETHERPAD_DB_PASSWORD}
      - DB_CHARSET=utf8mb4
      # Custom wait script env var
      - WAIT_FOR_SERVICES=etherpad_db:3306

```

See [docker-compose_mysql.yml](https://github.com/Monogramm/docker-etherpad/blob/master/template/docker-compose_mysql.yml) and [.env](https://github.com/Monogramm/docker-etherpad/blob/master/template/.env) for details.

The only thing that can be mentioned in our Dockerfile is that we added a "_wait for_" mechanism (inspired by [wait-for-it](https://github.com/vishnubob/wait-for-it)). It allows the container to wait for DB to be ready before starting :
- Wait for `host:port` to be ready and then execute command: [wait-for-service.sh](https://github.com/Monogramm/docker-etherpad/blob/b0a1e7ec722af86fe1a3604dc556250db41f73eb/template/wait-for-service.sh#L15)
- Change docker command to wait for service before starting etherpad-lite: [docker-compose_mysql.yml command](https://github.com/Monogramm/docker-etherpad/blob/b0a1e7ec722af86fe1a3604dc556250db41f73eb/template/docker-compose_mysql.yml#L31)
- Configure service to wait for through env var: [docker-compose_mysql.yml WAIT_FOR_SERVICES](https://github.com/Monogramm/docker-etherpad/blob/b0a1e7ec722af86fe1a3604dc556250db41f73eb/template/docker-compose_mysql.yml#L53)

The main difference with [wait-for-it](https://github.com/vishnubob/wait-for-it) is that we use env var to define wait and we do not need bash to be installed, only netcat (so smaller images).

**Comment by JohnMcLear** [comment]
Okay so yea do this: 

```
"SELECT CCSA.character_set_name FROM information_schema.`TABLES` T,information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation AND T.table_schema = 'DATABASE' AND T.table_name = 'store'";
```

replacing DATABASE with your database name.

The value should be ``utf8mb4``

**Comment by madmath03** [comment]
> The value should be utf8mb4.

I can confirm. This is what the `command` in the database are for.

Also, latest develop branch does fix the issue: https://travis-ci.org/github/Monogramm/docker-etherpad/jobs/669817706?utm_medium=notification&utm_source=github_status

**Comment by JohnMcLear** [comment]
<3 I'm gonna just leave this on my screen for a few days because I think I can refine this solution a little prior to 1.8.1
